README
Fork of Circular-Application-Menu (C-A-M) by Steffen Rumberger

A Circular-Application-Menu (C-A-M) mirroring the same structure as the existing application menu for the GNOME desktop.

It was originally just a paper prototype written by Cole Anstey. But his project is no longer maintained and no further contributions will be accepted by him. That's why i maintain a fork.

Cole Anstey's project website: http://circular-application-menu.googlecode.com

Other Implementations

	* http://gnome-pie.simonschneegans.de

REQUIREMENTS

	* GNOME platform and a composited environment (composition-manager or a composition-window-manager i.e. Compiz, Metacity, Mutter etc.)
	* GTK+ (GIMP Toolkit) Version 3.x
	* libgnome-menu Version 3.x

INSTALLATION

To install Circular-Application-Menu in your home directory run:

	git clone http://github.com/1nn3/circular-application-menu.git c-a-m
	cd c-a-m

	./autogen.sh # create configure script
	./configure --prefix=$HOME/.local
	make
	make install
	$HOME/.local/bin/cmm # run it

To build a package for Debian GNU/Linux run instead:

	apt-get install build-essential debhelper autotools-dev autoconf automake libgtk-3-dev libgnomevfs2-dev libgnome-menu-3-dev
	./autogen.sh # create configure script
	dpkg-buildpackage -b -uc

LICENSE AND COPYRIGHT

Copyright (C) 2008 - 2013 Cole Anstey <cole.anstey@gmail.com>
                          Steffen Rumberger <inne@sdfeu.org>

This program is free software. See COPYING and http://www.gnu.org/licenses/gpl.html for more information.

CONTACT

Your patches, help and feedback is greatly appreciated.

Contact me by email: Steffen Rumberger <inne@sdfeu.org>

THANKS

My thanks go to Cole Anstey, the project founder and author of Circular-Application-Menu and all people who have contributed. See AUTHORS for more information.



CHANGELOG
r57:
	* Added an option to colorable the menu.
	  Option -c, --color specifies colors to colorable the menu.
	* Added an option to hide the tooltip.
	  Option --hide-tooltip hides the tooltip. For all those, who wants the menu minimalism.
	  Fixed feature request issue 11.
	* Added an option to just maximize the window instead of place in the fullscreen state.
	  Option --no-fullscreen just maximize the window.
	  Usually the menu is a window in the fullscreen state, but maximized looks better with the GNOME-Shell.
	* Removed many short option names.
	* Migrated to libgnome-menu version 3.
	* Fixed an error where the menu is not correct displayed in multi-head envirements.
	  Center menu on main screen in multi-head environments.
	  Fixed issue 3.

r56 and older:
	* No changlog was written



src/main.h
#ifndef __MAIN_H__
#define __MAIN_H__

#define G_DISABLE_DEPRECATED
#define GSEAL_ENABLE
#define GTK_DISABLE_DEPRECATED
#define GTK_DISABLE_SINGLE_INCLUDES
#define GTK_MULTIHEAD_SAFE
#define GTK_MULTIDEVICE_SAFE
#define GDK_DISABLE_DEPRECATED
#define GDK_PIXBUF_DISABLE_DEPRECATED
#define GMENU_I_KNOW_THIS_IS_UNSTABLE

#include <gtk/gtk.h>
#include <gdk/gdkkeysyms.h>
#include <gdk/gdkx.h>
#include <gio/gdesktopappinfo.h>
#include <libgnomevfs/gnome-vfs.h>
#include <X11/Xlib.h>
#include <X11/Xatom.h>
#include <glib/gi18n.h>
#include <gnome-menus-3.0/gmenu-tree.h>

#include <math.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>

#include "../config.h"
#include "cmmcircularmainmenu.h"

#include "../pixmaps/close-menu-prelight.h"
#include "../pixmaps/close-menu-normal.h"
#include "../pixmaps/open-sub-menu-prelight.h"
#include "../pixmaps/open-sub-menu-normal.h"
#include "../pixmaps/reflection.h"

#endif


src/cmmcircularmainmenu.h
/* circular-main-menu
 * Copyright: 2008-2013 Cole Anstey <cole.anstey@gmail.com>
 *                      Steffen Rumberger <inne@sdfeu.org>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 *
 * See http://github.com/1nn3/circular-application-menu
 */
#ifndef __CA_CIRCULAR_APPLICATION_MENU_H__
#define __CA_CIRCULAR_APPLICATION_MENU_H__

#include "main.h"

G_BEGIN_DECLS

#define CA_TYPE_CIRCULAR_APPLICATION_MENU          (ca_circular_application_menu_get_type ())
#define CA_CIRCULAR_APPLICATION_MENU(obj)          (G_TYPE_CHECK_INSTANCE_CAST ((obj), CA_TYPE_CIRCULAR_APPLICATION_MENU, CaCircularApplicationMenu))
#define CA_CIRCULAR_APPLICATION_MENU_CLASS(obj)    (G_TYPE_CHECK_CLASS_CAST ((obj), CA_CIRCULAR_APPLICATION_MENU, CaCircularApplicationMenuClass))
#define CA_IS_CIRCULAR_APPLICATION_MENU(obj)       (G_TYPE_CHECK_INSTANCE_TYPE ((obj), CA_TYPE_CIRCULAR_APPLICATION_MENU))
#define CA_IS_CIRCULAR_APPLICATION_MENU_CLASS(obj) (G_TYPE_CHECK_CLASS_TYPE ((obj), EFF_TYPE_CIRCULAR_APPLICATION_MENU))
#define CA_CIRCULAR_APPLICATION_MENU_GET_CLASS     (G_TYPE_INSTANCE_GET_CLASS ((obj), CA_TYPE_CIRCULAR_APPLICATION_MENU, CaCircularApplicationMenuClass))

/* Leaf types. */
typedef enum
{
    ROOT_LEAF,
    FILE_LEAF,
    SUB_FILE_LEAF
} LeafType;

/* Gylph types. */
typedef enum
{
    GLYPH_UNKNOWN,
    GLYPH_ROOT_CENTRE,
    GLYPH_FILE_MENU_CENTRE,
    GLYPH_FILE,
    GLYPH_FILE_MENU,
    GLYPH_TAB
} GlyphType;

/* The current animation state of a leaf. */
typedef enum
{
    NO_STATE,
    INITIAL_OPEN
} AnimationState;

typedef struct _CaCircularApplicationMenu        CaCircularApplicationMenu;
typedef struct _CaCircularApplicationMenuClass   CaCircularApplicationMenuClass;

struct _CaCircularApplicationMenu
{
    GtkDrawingArea parent;

    /*< private >*/

    /*< public >*/
};

struct _CaCircularApplicationMenuClass
{
    GtkDrawingAreaClass parent_class;
};

typedef struct _CaMenuRender CaMenuRender;
typedef struct _CaSegmentRender CaSegmentRender;
typedef struct _CaFileItem CaFileItem;
typedef struct _CaFileLeaf CaFileLeaf;

/* Structure to hold calculated menu values for rendering. */
/* Held here to stop them being calculated more than once. */
struct _CaMenuRender
{
    // Render a child menu.
    /*
     C   E
     B   F
    A     A
    A     A
    A     A
     AAAAA
    */
    gint Bcircle_x;
    gint Bcircle_y;
    gint Ccircle_x;
    gint Ccircle_y;
    gint Ecircle_x;
    gint Ecircle_y;
    gint Fcircle_x;
    gint Fcircle_y;
    gint Bcircle_arc_x;
    gint Bcircle_arc_y;
    gint Ccircle_arc_x;
    gint Ccircle_arc_y;
    gdouble lowest_angle_nearest;
    gdouble highest_angle_nearest;
    gdouble Bcircle_lowest_angle;
    gdouble Bcircle_highest_angle;
    gdouble Fcircle_lowest_angle;
    gdouble Fcircle_highest_angle;
    gdouble lowest_angle_farthest;
    gdouble highest_angle_farthest;
    gdouble Ccircle_lowest_angle;
    gdouble Ccircle_highest_angle;
    gdouble Ecircle_lowest_angle;
    gdouble Ecircle_highest_angle;

    gdouble tab_glyph_radius;
    gint tab_glyph_x;
    gint tab_glyph_y;
};

/* Structure to hold calculated segment values for rendering. */
/* Held here to stop them being calculated more than once.    */
struct _CaSegmentRender
{
    /*
    Segment

    CDC
    BAB
    */
    gint Aradius;
    gint Bradius;
    gint Cradius;
    gint Dradius;
    gdouble Afrom_angle;
    gdouble Bfrom_angle;
    gdouble Cfrom_angle;
    gdouble Dfrom_angle;
    gdouble Ato_angle;
    gdouble Bto_angle;
    gdouble Cto_angle;
    gdouble Dto_angle;
    gint B_lowest_circle_x;
    gint B_lowest_circle_y;
    gint C_lowest_circle_x;
    gint C_lowest_circle_y;
    gint B_highest_circle_x;
    gint B_highest_circle_y;
    gint C_highest_circle_x;
    gint C_highest_circle_y;

    // Arrow.
    gdouble arrow_percentage;
    gdouble common_angle;
    gint arrow_point_x;
    gint arrow_point_y;
    gint arrow_side_x;
    gint arrow_side_y;
    gint arrow_radius;
    gdouble arrow_angle;
};

/* The file-item which resides on a file-leaf. */
struct _CaFileItem
{
    GlyphType _type;
    gpointer _menutreeitem; /* GMenuTreeDirectory* */
    const gchar* _icon;
    GdkPixbuf* _pixbuf;
    CaFileLeaf* _assigned_fileleaf;
    CaFileItem* _associated_fileitem;
    gdouble _parent_angle;
    gdouble _circular_angle_share;
    gdouble _parent_radius;
    CaSegmentRender* _segment_render;

    /*< Drawing >*/
    gint x;
    gint y;
    gint size;
};

/* The file-leaf containing 0-* file-items. */
struct _CaFileLeaf
{
    LeafType _type;
    GList* _fileitem_list;
    gint _fileitem_list_count;
    AnimationState _current_animation_state;
    gint _current_animation_tick;
    CaFileLeaf* _parent_fileleaf;
    CaFileLeaf* _child_fileleaf;
    GList* _sub_fileleaves_list;
    CaFileItem* _central_glyph;
    CaMenuRender* _menu_render;

    /*< Drawing >*/
    gint x;
    gint y;
    gint radius;
};

/* Construction */
GtkWidget* ca_circular_application_menu_new (GdkWindow *window, gboolean hide_preview, gboolean hide_tooltip, gboolean warp_mouse, gint glyph_size, gchar* emblem, gboolean render_reflection, gboolean render_tabbed_only, gchar* color);
GType ca_circular_application_menu_get_type(void);

/* Implementation */
CaFileLeaf* ca_circular_application_menu(CaCircularApplicationMenu* circular_application_menu, GMenuTreeDirectory* menutreedirectory);
CaFileLeaf* ca_circular_application_menu_show_leaf(CaCircularApplicationMenu* circular_application_menu, GMenuTreeDirectory* menutreedirectory, LeafType leaftype, CaFileItem* fileitem, gboolean disassociated);
void ca_circular_application_menu_close_fileleaf(CaFileLeaf* fileleaf);

/* Private functions. */

G_END_DECLS

#endif


src/main.c
/* circular-main-menu
 * Copyright: 2008-2013 Cole Anstey <cole.anstey@gmail.com>
 *                      Steffen Rumberger <inne@sdfeu.org>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 *
 * See http://github.com/1nn3/circular-application-menu
 */

#include "main.h"

static CaFileLeaf* root_fileleaf = NULL;

/**
 * _ca_circular_application_menu_enable_blur:
 * @widget: The widget to set the blur against.
 *
 * Uses the compiz blur plugin to blur underneath the menus which makes the rendering clearer.
 **/
static void
_ca_circular_application_menu_enable_blur(GtkWidget* window)
{
    Display* xdisplay;
    long data[2];

    /* Blur under the window.  Taken from the blur demo application. */
    xdisplay = GDK_DISPLAY_XDISPLAY (gdk_display_get_default ());

    data[0] = 9;    /* threshold */
    data[1] = 0;    /* filter */

    XChangeProperty (
        xdisplay,
        GDK_WINDOW_XID (gtk_widget_get_window (window)),
        XInternAtom (xdisplay, "_COMPIZ_WM_WINDOW_BLUR", FALSE),
        XA_INTEGER,
        32, PropModeReplace, (guchar *) data,
        2);
}

int
main (int argc, char **argv)
{
    GtkWidget* window;
    GdkScreen* screen;
    GdkVisual* visual;
    GtkWidget* circular_application_menu;
    GOptionContext* optioncontext;
    GError* error = NULL;
    GMenuTreeFlags flags;
    GMenuTree* tree;
    GMenuTreeDirectory* root;
    /* Default values. */
    gboolean hide_preview = FALSE;
    gboolean warp_pointer_off = FALSE;
    gint glyph_size = 3;
    gboolean blur_off = FALSE;
    gchar* emblem = NULL;
    gboolean render_reflection = FALSE;
    gboolean render_tabbed_only = FALSE;
    gchar *menu_file = NULL;
    /*gboolean  monitor = FALSE;*/
    gboolean include_excluded = FALSE;
    gboolean include_nodisplay = FALSE;
    gboolean hide_tooltip = FALSE;
    gchar* color = NULL;
    gboolean no_fullscreen = FALSE;
    /* Key-value file parser — parses .ini-like config files */
    GKeyFile * key_file = g_key_file_new ();
    gchar key_file_path[30] = "";
    g_sprintf(&key_file_path, "%s/.cmm\0", getenv("HOME"));
    if ( g_key_file_load_from_file (key_file, &key_file_path, G_KEY_FILE_NONE, &error))
    {
        color = g_key_file_get_string(key_file, "DEFAULT", "color", NULL);
        emblem = g_key_file_get_string(key_file, "DEFAULT", "emblem", NULL);
        menu_file = g_key_file_get_string(key_file, "DEFAULT", "menu-file", NULL);
        glyph_size = g_key_file_get_integer(key_file, "DEFAULT", "glyph-size", NULL);
        no_fullscreen = g_key_file_get_boolean(key_file, "DEFAULT", "no-fullscreen", NULL);
        include_excluded = g_key_file_get_boolean(key_file, "DEFAULT", "include-excluded", NULL);
        include_nodisplay = g_key_file_get_boolean(key_file, "DEFAULT", "include-nodisplay", NULL);
        hide_tooltip = g_key_file_get_boolean(key_file, "DEFAULT", "hide-tooltip", NULL);
        hide_preview = g_key_file_get_boolean(key_file, "DEFAULT", "hide-preview", NULL);
        warp_pointer_off = g_key_file_get_boolean(key_file, "DEFAULT", "warp-pointer-off", NULL);
        blur_off = g_key_file_get_boolean(key_file, "DEFAULT", "blur-off", NULL);
        render_reflection = g_key_file_get_boolean(key_file, "DEFAULT", "render-reflection", NULL);
        render_tabbed_only = g_key_file_get_boolean(key_file, "DEFAULT", "render-tabbed-only", NULL);
    }
    else if (error)
    {
        g_message ("Failed to load %s: %s", key_file_path, error->message);
        g_clear_error (&error); /* g_error_free */
    }

    GOptionEntry options[] =
    {
        { "color", 'c', 0, G_OPTION_ARG_STRING, &color, _("Specifies the colors (colon separated) to use for segments and text."), "NORMAL[:PRELIGHT[:TEXT]]" },
        { "hide-preview", '\0', 0, G_OPTION_ARG_NONE, &hide_preview, "Hides the menu preview displayed when the mouse is over a menu.", NULL },
        { "hide-tooltip", '\0', 0, G_OPTION_ARG_NONE, &hide_tooltip, _("Hides the tooltip displayed when the mouse is over a menuitem."), NULL },
        { "warp-pointer-off", '\0', 0, G_OPTION_ARG_NONE, &warp_pointer_off, "Stops the pointer e.g. the mouse from warping to the centre of the screen whenever a menu is shown.", NULL },
        { "glyph-size", 'g', 0, G_OPTION_ARG_INT, &glyph_size, "The size of the glyphs [SIZE: 1=small, 2=medium or 3=large (default)]. ", "SIZE"  },
        { "blur-off", '\0', 0, G_OPTION_ARG_NONE, &blur_off, "Stops the blur from underneath the menu.", NULL },
        { "emblem", 'e', 0, G_OPTION_ARG_STRING, &emblem, "Specifies the (colon separated) emblems to use for the root menu.", "NORMAL:PRELIGHT" },
        { "render-reflection", '\0', 0, G_OPTION_ARG_NONE, &render_reflection, "Stops the reflection from being rendered.", NULL },
        { "render-tabbed-only", '\0', 0, G_OPTION_ARG_NONE, &render_tabbed_only, "Only renders the currently tabbed menu.", NULL },        
        { "menu-file", '\0', 0, G_OPTION_ARG_STRING, &menu_file, N_("Menu file."), N_("MENU_FILE") },
        { "no-fullscreen", '\0', 0, G_OPTION_ARG_NONE, &no_fullscreen, "Maximize the window instead of place in the fullscreen state.", NULL },
        { "include-excluded", '\0', 0, G_OPTION_ARG_NONE, &include_excluded, N_("Include excluded menu entries."), NULL },
        { "include-nodisplay", '\0', 0, G_OPTION_ARG_NONE, &include_nodisplay, N_("Include NoDisplay=true menu entries."), NULL },
        { NULL }
    };

    /* Initialise. */
    gtk_init (&argc, &argv);
    gnome_vfs_init();

    /* Parse the arguments. */
    optioncontext = g_option_context_new(
        "- Circular-Main-Menu a Circular-Application-Menu (C-A-M) representation for the GNOME desktop");
    g_option_context_add_main_entries(optioncontext, options, "circular_applications");
    g_option_context_parse(optioncontext, &argc, &argv, NULL);

    if (!g_option_context_parse (optioncontext, &argc, &argv, &error))
    {
        g_critical (_("Option parsing failed: %s"), error->message);
        abort();
    }

    g_option_context_free(optioncontext);

    flags = GMENU_TREE_FLAGS_NONE;
    if (include_excluded)
        flags |= GMENU_TREE_FLAGS_INCLUDE_EXCLUDED;
    if (include_nodisplay)
        flags |= GMENU_TREE_FLAGS_INCLUDE_NODISPLAY;
    
    tree = gmenu_tree_new (menu_file ? menu_file : "applications.menu", flags);
    g_assert (tree != NULL);

    if (!gmenu_tree_load_sync (tree, &error))
    {
        g_error ("Failed to load tree: %s", error->message); /* resulting in a call to abort() */
        /* abort(); */
    }

    g_message ("Loaded menu from %s", gmenu_tree_get_canonical_menu_path (tree));

    root = gmenu_tree_get_root_directory (tree);

    if (root == NULL)
    {
        g_warning (_("The menu tree is empty."));
    }

    if (FALSE == gdk_display_supports_composite(gdk_display_get_default()))
    {
        g_warning(_("The circular-main-menu only displays correctly with composited desktops."));
    }
    screen = gdk_screen_get_default ();

    window = gtk_window_new (GTK_WINDOW_TOPLEVEL);
    /* Close the window when it lose the focus. */
    gtk_widget_set_events (window, GDK_FOCUS_CHANGE_MASK);
    g_signal_connect (G_OBJECT(window),"focus-out-event", G_CALLBACK (gtk_main_quit), NULL);

    /* Assign an alpha colormap to the window. */
    screen = gtk_widget_get_screen (GTK_WIDGET (window)); // screen = gdk_screen_get_default ();
    visual = gdk_screen_get_rgba_visual (screen);

    if (visual == NULL)
        visual = gdk_screen_get_system_visual (screen);

    gtk_widget_set_visual (window, visual);
    /*
    Does not seem to be required.
    gdk_window_set_decorations(window->window, 0);
    */
    gtk_widget_realize (window);    /* This is required otherwise gdk_pixmap_new() will assert. */

    /* Constructs a new dockband widget. */
    circular_application_menu = ca_circular_application_menu_new (
        gtk_widget_get_window (GTK_WIDGET (window)),
        hide_preview,
        hide_tooltip,
        warp_pointer_off,
        glyph_size,
        emblem,
        render_reflection,
        render_tabbed_only,
        color);
    gtk_container_add (GTK_CONTAINER (window), circular_application_menu);

    g_signal_connect (window, "destroy", G_CALLBACK (gtk_main_quit), NULL);

    /* Make the application full screen, without this it will be below any top edged panel. */
    if (!no_fullscreen) {
        gtk_window_fullscreen(GTK_WINDOW(window));
    } else {
        gdk_window_set_decorations(gtk_widget_get_window (GTK_WIDGET (window)), 0);
        gtk_window_maximize(GTK_WINDOW(window));
    }

    /* Check whether blur is enabled. */
    if (FALSE == blur_off)
    {
        /* Use the compiz blur plugin to blur underneath the menus which makes the rendering clearer. */
        _ca_circular_application_menu_enable_blur(window);
    }

    gtk_widget_show_all (window);

    /* Shows the menu tree directory which becomes the root file leaf. */
    root_fileleaf = ca_circular_application_menu(CA_CIRCULAR_APPLICATION_MENU(circular_application_menu), root);

    /* Invalidate the widget. */
    gtk_widget_queue_draw(window);

    gtk_main ();

    gmenu_tree_item_unref (root);

    return 0;
}
